### ISSUE TITLE
Randomly fails with Promise all style where 10+ hosts are pinged at once and cannot ping 50+ at the same time

### ISSUE THREAD

**Issue by ernestgwilsonii** [stephenwvickers/node-net-ping#47]
I am using Promise all because I want to ping ~400 hosts at a time.
It seems like trying to ping ~10 hosts (or more) randomly fails.
I have never been able to ping say even 50+ at the same time ever.
Does ping or raw socket have some concurrent ping limitations or settings?

C:\>node --version
v8.9.3

Promise all and async-await (anything that tries to ping multiple in parallel) fails with high numbers of simultaneous pings:
https://github.com/ernestgwilsonii/javascript-ping-examples/tree/master/examples

**Comment by ernestgwilsonii** [comment]
I tried on Linux  too, same issue.

I even tried an old school callback style running these pings in a batch of 70 sequentially but fast, boom:
$ node index5.js
70
Error: -4071
Assertion failed: !(handle->flags & UV__HANDLE_CLOSING), file src\win\poll.c, line 475

At least we get some type of error when trying to run a bunch of pings using callback style.
Promise methods seem to fail silently (annoying).

I really want to use a FAST binary and native Node.js to PING many hosts and not be forced to shell out to the OS ping via child-process.
What can be done?

**Comment by stephenwvickers** [comment]
Hi, I am not aware of any such limitations.  There is no Promise support built into the library so I assume you are wrapping it with code, or using another module which does.

Can you please provide a full program that runs and emits the issue, I will can then look at recreating the issue in my lab?

Thanks

**Comment by ernestgwilsonii** [comment]
Hi Stephen, here are two examples for testing:
https://github.com/ernestgwilsonii/temp-repo-for-node-net-ping-issue47

One is old school callback style, the other is newer promise all style.
Both suffer from the same issues:
- either failing silently part way through a large list of things to ping
or
- sometimes, but rarely an error come in as:
Error: -4071

This happens on Windows and Linux.
Small lists PING and complete OK, bigger lists have issues.
Does ping or raw socket have some concurrent ping limitations or settings?

Thanks,
Ernest

**Comment by stephenwvickers** [comment]
Hi @ernestgwilsonii 

The library works fine for me, and while I do see your issue with both your programs I don't think your issue is net-ping.

If you perform a Wireshark capture during the ping requests you will in fact see that when net-ping times out no response is indeed received for many ping requests.

During the ping runs many pings are sent to specific IP addresses, do you not think these hosts, or some network infrastructure along the path, is ignoring your pings because you are performing too many than what is permitted?

Currently, from what I can see, there is no issue with net-ping.

Steve

**Comment by ernestgwilsonii** [comment]
Steve, thanks for looking!
Itt should not matter whether or not PING responds.
My point is, when net-ping fails, you never reach line 99
in ping_many_callback-style-all-in-one.js
Attempting to ping large numbers of devices seems to cause net-ping to fail (silently in most cases)
and when it does, the calling script silently dies with it.
Since my methods of calling net-ping may be contributing to confusion...
Can you please supply a simple example script where you ping say 100+ targets in parallel?
I suspect you may see the issue I am reporting manifest itself in some form.
Thanks Steve!

**Comment by somec** [comment]
Hi,
I was looking at ernest code and noticed that you call ping.createSession in the execute function, so each of your hosts has their own session. That's not how it is supposed to work. You are supposed to have one global session (or maybe 2 if you want both ipv4 and ipv6).

To test, I moved: const session = ping.createSession(this.options);
to right after your declarations of global options and changed it from const to var.

After that it works as expected on windows with node 8.9.3 and on linux with 9.3.0.

**Comment by ernestgwilsonii** [comment]
Thank you Steve!
Just to be certain that I test correctly can you upload or paste or attach the fully working file.js and I will give it a try.
Thanks again!

**Comment by somec** [comment]
Hi @ernestgwilsonii

I made a pull request to your examples. 
When I made my comment 2 days ago I had only looked at ping_promise_promise-all_all-in-one.js, where you use a promise.

This time I also looked at ping_many_callback-style-all-in-one.js, and now I see you have more issues with your code. For example:

```
pingChunker(targets, (err, res) => {
    //
});

console.log(targets.length); // When it fails, the script silently exits and this line never prints
```
In **pingChunker** you make a lot of calls that requires a callback. But you never wait for the callbacks to return. Therefore your console.log will not execute when you expect it to.

```
function pingChunker(arr, callback) {
    for (let target of arr) {
        // Callback style!
        pinger(target, options, (err, res) => {
          // [code removed]
          // this code will may be executed sometime in the future, when the callback is called from pinger.
        });
    }
    callback(); // This callback is called when the for loop above is finished, but before any callbacks has been called.
}
```

**Comment by stephenwvickers** [comment]
Hi @ernestgwilsonii 

See the comment from @somec.  It appears there are some issues with your script, and some suggestions have been given.

Unfortunately, I genuinely don't have the time to update and test your code, I have a lot of work on right now, and other module enhancements to get done.

I'll mark this as closed and if you find another issue please open another ticket.

@somec Thanks for your efforts on this one.

Steve

**Comment by ernestgwilsonii** [comment]
@stephenwvickers  and @somec Thank you both!
To put it bluntly...
I was doing it wrong! (facepalm sorry)
Please do consider this closed!
Thanks for responding!

**Comment by stephenwvickers** [comment]
:) well, thanks for your honesty!

Hope you find some use for it.

Steve
